File: WOA_TSP.py
import random
import numpy as np
import math
from matplotlib import pyplot as plt

class City(object):  # 城市类
    def __init__(self, ID, CoordinateX, CoordinateY):
        self.ID = ID
        self.CoordinateX = CoordinateX
        self.CoordinateY = CoordinateY

    def shuchu(self):
        print(self.ID, self.CoordinateX, self.CoordinateY)

def getDistanceOfTwoCity(city1, city2):  # 计算两个城市距离的函数
    return math.sqrt((pow((city1.CoordinateX - city2.CoordinateX), 2) + pow((city1.CoordinateY - city2.CoordinateY), 2)) / 10.0)

def ReplaceContinueousSpace(str): # 替换字符串中的空格，并以逗号分隔符连接
    n = len(str)
    newstr = ''
    for i in range(n):
        if str[i] != ' ':
            newstr += str[i]
            if i+1 < n and str[i+1] == ' ':
                newstr += ','
    return newstr


Distances = []  # 城市距离矩阵

class CityGraph(object):  # 城市群类
    def __init__(self, filename):
        self.Citys = []  # 所有的城市
        self.n = 0  # 城市的数量
        self.readDataFile(filename)
        self.computeDistances()

    def readDataFile(self, filename):  # 读取城市数据文件函数
        f = open(filename)
        line = f.readline()
        while line:
            line_str = line.strip('\n')
            city_linestr = ReplaceContinueousSpace(line_str).split(',')
            city_i = City(int(city_linestr[0]), float(city_linestr[1]), float(city_linestr[2]))
            self.Citys.append(city_i)
            line = f.readline()
        f.close()
        self.n = len(self.Citys)

    def computeDistances(self):  # 得到城市距离矩阵的函数
        for i in range(0, self.n):
            dist_i = []
            for j in range(0, self.n):
                dist_ij = getDistanceOfTwoCity(self.Citys[i], self.Citys[j])
                dist_i.append(dist_ij)
            Distances.append(dist_i)

    def shuchu(self):  # 输出城市群信息的函数
        print('城市ID-----城市横坐标X-----城市纵坐标Y')
        for i in range(self.n):
            self.Citys[i].shuchu()
        print('----------各城市间的距离矩阵---------')
        for i in range(0,self.n):
            for j in range(0, self.n):
                print(round(Distances[i][j],3), end=',')
            print('\n')

# The Whale Optimization Algorithm By jing_zhong 2022.6.2
class WOA(object):
    def __init__(self, population_size=50, dimension=29, lb=1, ub=29, Max_iteration = 100):
        self.dimension = dimension  # 每个鲸鱼个体位置的维度
        self.lb = lb  # 鲸鱼个体位置的下边界
        self.ub = ub  # 鲸鱼个体位置的上边界
        self.PopSize = population_size  # 鲸鱼种群大小
        self.Max_iter = Max_iteration  # 最大迭代次数
        self.Population = []  # 鲸鱼种群
        self.BestGeti = []  # 最优鲸鱼个体
        self.BestFitness = -100  # 最优鲸鱼个体的适应度
        self.Convergence_curve = []  # 各代鲸鱼种群最优适应度集合,用于绘制收敛曲线

    def Init_Population(self):  # 种群初始化，评价所有鲸鱼个体的适应度，获得最优个体
        self.Population = []
        for i in range(0, self.PopSize):
            geti_i = []
            for j in range(0, self.dimension):
                xj = j + 1
                geti_i.append(xj)
            random.shuffle(geti_i)
            self.Population.append(geti_i)
        self.BestGeti = self.Population[0]
        self.BestFitness = self.Evaluate(self.Population[0])
        for i in range(0,self.PopSize):
            fitness_i = self.Evaluate(self.Population[i])
            print('种群初始化第 {} 个服务链座鱼鲸个体 {} 的适应度为 {}'.format(i + 1, self.Population[i], fitness_i))
            if fitness_i < self.BestFitness:
                self.BestGeti = self.Population[i]
                self.BestFitness = fitness_i
        print('\n初始化种群最优服务链座鱼鲸个体 {} 的适应度为 {}'.format(self.BestGeti, self.BestFitness))

    def Valid_Geti(self,geti):
        for j in range(0, self.dimension):
            if geti[j] > self.lb and geti[j] < self.ub:
                geti[j] = int(math.floor(geti[j]))
            else:
                geti[j] = random.randint(int(self.lb), int(self.ub))
        # 调整鲸鱼个体解的有效性
        n = len(geti)
        std_geti = []  # 标准的一个全排列
        std_flag = []  # 标记是否需要修改
        Count = []  # 统计出现的次数
        for i in range(0, n):
            std_geti.append(i + 1)
            Count.append(0)
            std_flag.append(True)
        for i in range(0, n):
            findNum = std_geti[i]
            count = 0
            for j in range(0, n):
                if (findNum == geti[j]):
                    count = count + 1
                    if count > 1:
                        std_flag[j] = False
            Count[i] = count
        for i in range(0, n):
            if Count[i] == 0:
                for j in range(0, n):
                    if std_flag[j] == False:
                        geti[j] = std_geti[i]
                        std_flag[j] = True
                        break

    def Evaluate(self, geti):  # 鲸鱼个体适应度评价函数
        sumDist = 0
        n = len(geti)
        for i in range(0, n-1):
            sumDist += Distances[geti[i]-1][geti[i+1]-1]
        sumDist += Distances[geti[n-1]-1][geti[0]-1]
        return int(sumDist)

    def Start(self):
        self.Init_Population() # 初始化鲸鱼种群
        t = 0
        while t < self.Max_iter:
            a = 2 - t * (2 / self.Max_iter)  # a decreases linearly fron 2 to 0
            a2 = -1 + t * ((-1) / self.Max_iter)  # a2 linearly dicreases from -1 to -2 to calculate t
            for i in range(0, self.PopSize):
                r1 = random.random()  # r1 is a random number in [0,1]
                r2 = random.random()  # r2 is a random number in [0,1]
                A = 2 * a * r1 - a
                C = 2 * r2
                b = 1
                l = (a2 - 1) * random.random() +1
                p = random.random()
                for j in range(0,self.dimension):
                    if p < 0.5:
                        if abs(A) >= 1:
                            rand_leader_index = math.floor(self.PopSize * random.random())
                            X_rand = self.Population[rand_leader_index]
                            D_X_rand = abs(C * X_rand[j] - self.Population[i][j])
                            self.Population[i][j] = X_rand[j] - A * D_X_rand
                        else:
                            D_Leader = abs(C * self.BestGeti[j] - self.Population[i][j])
                            self.Population[i][j] = self.BestGeti[j] - A * D_Leader
                    else:
                        distance2Leader = abs(self.BestGeti[j] - self.Population[i][j])
                        self.Population[i][j] = distance2Leader * np.exp(b * l) * np.cos(l * 2 * math.pi) + self.BestGeti[j]
            t = t + 1
            self.Convergence_curve.append(self.BestFitness)
            for i in range(0,self.PopSize):
                self.Valid_Geti(self.Population[i]) # Return back the search agents that go beyond the boundaries of the search space
                fitness_i = self.Evaluate(self.Population[i])
                if fitness_i < self.BestFitness:
                    self.BestGeti = self.Population[i]
                    self.BestFitness = fitness_i
                print('第 {} 代种群第 {} 个座鱼鲸个体 {} 的适应度为 {}'.format(t + 1, i+1, self.Population[i], fitness_i))
            print('第 {} 代种群 最优座鱼鲸个体 {} 的适应度为 {}'.format(t + 1, self.BestGeti, self.BestFitness))
        print('{} 次迭代后WOA种群最优座鱼鲸个体 {} 的适应度值为 {}'.format(self.Max_iter, self.BestGeti, self.BestFitness))

if __name__ == "__main__":
    filename = 'bayg29.tsp'
    citygraph = CityGraph(filename)
    citygraph.shuchu()
    woa = WOA(50, 29, 1, 29, 500)
    woa.Start()
    epochs = range(len(woa.Convergence_curve))
    plt.figure()
    plt.plot(epochs, woa.Convergence_curve, "b", label="woa")
    plt.title('WOA_ServiceChain')
    plt.xlabel("Epochs")
    plt.ylabel("Fitness")
    plt.legend()
    plt.show()

File: WOA_TSP.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <random>
#include <ctime>

using namespace std;

const int citycount = 29;//城市的数量
#define PI       3.14159265358979323846   // pi
#define N  999 //精度为小数点后面3位
double round(double r)
{
    return (r > 0.0) ? floor(r + 0.5) : ceil(r - 0.5);
}

class City//城市类
{
    public:
        string ID;//城市名称ID
        double x, y;//城市点的二维坐标x和y
        void shuchu()
        {
            std::cout <<ID+":"<<"("<< x << "," << y <<")"<< endl;
        }
};

class CityGraph//城市图
{
    public:
        City city[citycount];//城市数组
        double distance[citycount][citycount];//城市间的距离矩阵
        void Readcoordinatetxt(string txtfilename)//读取城市坐标文件的函数
        {
            ifstream myfile(txtfilename, ios::in);
            double x = 0, y = 0;
            int z = 0;
            if (!myfile.fail())
            {
                int i = 0;
                while (!myfile.eof() && (myfile >> z >> x >> y))
                {
                    city[i].ID = to_string(z);//城市名称转化为字符串
                    city[i].x = x; city[i].y = y;
                    i++;
                }
            }
            else
                cout << "文件不存在";
            myfile.close();//计算城市距离矩阵
            for (int i = 0; i < citycount; i++)
                for (int j = 0; j < citycount; j++)
                {
                    distance[i][j] = sqrt((pow((city[i].x - city[j].x), 2) + pow((city[i].y - city[j].y), 2))/10.0);//计算城市ij之间的伪欧式距离
                    if (round(distance[i][j] < distance[i][j]))distance[i][j] = round(distance[i][j]) + 1;
                    else distance[i][j] = round(distance[i][j]);
                }
        }
        void shuchu()
        {
            cout << "城市名称 " << "坐标x" << " " << "坐标y" << endl;
            for (int i = 0; i < citycount; i++)
                city[i].shuchu();
            cout << "距离矩阵： " << endl;
            for (int i = 0; i < citycount; i++)
            {
                for (int j = 0; j < citycount; j++)
                {
                    if (j == citycount - 1)
                       cout << distance[i][j] << endl;
                    else
                        cout << distance[i][j] << "  ";
                }
            }
        }
};


CityGraph Map_City;//全局对象城市图

int * Random_N(int n)// 随机生成一个个体解的函数
{
	int *geti;
	geti = new int[n];
	int j = 0;
	while(j<n)
	{
		while (true)
		{
			int flag = -1;
			int temp = rand() % n + 1;
			if (j > 0)
			{
				int k = 0;
				for(; k < j; k++)
				{
					if (temp == *(geti + k))break;
				}
				if (k == j)
				{
					*(geti + j) = temp;
					flag = 1;
				}
			}
			else
			{
				*(geti + j) = temp;
				flag = 1;
			}
			if (flag == 1)break;
		}
		j++;
	}
	return geti;
}

class JingYu//鲸鱼类，一个座头鲸代表一个个体解
{
	public:
		int *X;//鲸鱼位置向量
		double fitness;//鲸鱼的适应度
		void Init()
		{
			X = new int[citycount];
			int *M = Random_N(citycount);
			for (int j = 0; j < citycount; j++)
				X[j] = *(M + j);
			fitness=0;
		}
		void shuchu()
		{
			for (int j = 0; j < citycount; j++)
			{
				if (j == citycount -1) std::cout << X[j] << " "<<fitness<<endl;
				else std::cout << X[j] << "->";
			}
		}
};
class WOA
{
    public:
        JingYu *Population;//下一代种群
        int dimension;
        double LB;//位置向量每个分量的下确界
        double UB;//位置向量每个分量的上确界
        int Pop_Size;//种群大小
        int Itetime;//迭代次数
        JingYu Bestgeti;//最优鲸鱼个体
        double BestFitness;//最优鲸鱼个体的适应度


        void Init(int popsize,double dimen,double lb,double ub,int itetime,string filename)
        {
            Map_City.Readcoordinatetxt(filename);
            Map_City.shuchu();
            Pop_Size = popsize;
            dimension = dimen;
            LB = lb;
            UB = ub;
            Itetime = itetime;
            Population = new JingYu[Pop_Size];
            for (int i = 0; i < Pop_Size; i++)
            {
                Population[i].Init();
                Population[i].fitness = Evaluate(Population[i]);
            }
            Bestgeti.Init();
            for (int j = 0; j < citycount; j++)
                Bestgeti.X[j] = Population[0].X[j];
            Bestgeti.fitness = Evaluate(Bestgeti);
            for (int i = 0; i < Pop_Size; i++)
            {
                if (Population[i].fitness < Bestgeti.fitness)
                {
                    for (int j = 0; j < citycount; j++)
                        Bestgeti.X[j] = Population[i].X[j];
                    Bestgeti.fitness = Evaluate(Bestgeti);
                }
            }
            cout<<"初始化鲸鱼种群如下："<<endl;
            for (int i = 0; i < Pop_Size; i++)
                Population[i].shuchu();
            cout<<"初始化最优的鲸鱼个体如下："<<endl;
            Bestgeti.shuchu();
        }

        void Adjuxt_validGeti(JingYu jy)//调整鲸鱼个体位置有效性的函数
        {
            for(int j=0;j<dimension;j++)
            {

                if(jy.X[j] > LB && jy.X[j] < UB)
                    jy.X[j] = (int)jy.X[j];
                else
                    jy.X[j] = int(rand() % dimension);
            }
            int route[citycount];
            bool flag[citycount];
            int biaoji[citycount];
            for (int j = 0; j < citycount; j++)
            {
                route[j] = j + 1;
                flag[j] = false;
                biaoji[j] = 0;
            }

            for (int j = 0; j < citycount; j++)
            {
                int num = 0;
                for (int k = 0; k < citycount; k++)
                {
                    if (jy.X[k] == route[j])
                    {
                        biaoji[k] = 1;
                        num++; break;
                    }
                }
                if (num == 0) flag[j] = false;
                else if (num == 1) flag[j] = true;
            }
            for (int k = 0; k < citycount; k++)
            {
                if (flag[k] == false)
                {
                    int i = 0;
                    for (; i < citycount; i++)
                    {
                        if (biaoji[i] != 1)break;
                    }
                    jy.X[i] = route[k];
                    biaoji[i] = 1;
                }
            }
        }
        double Evaluate(JingYu jy)//计算鲸鱼个体适应值的函数
        {
            double Sum_dist=0;
            for (int j = 0; j < citycount-1; j++)
                Sum_dist += Map_City.distance[jy.X[j]][jy.X[j + 1]];
            Sum_dist += Map_City.distance[jy.X[citycount -1]][jy.X[0]];
            jy.fitness = Sum_dist;
            return Sum_dist;
        }

        void UpdateBestgeti()//更新鲸鱼最优个体的函数
        {
            for (int i = 0; i < Pop_Size; i++)
            {
                if (Population[i].fitness < Bestgeti.fitness)
                {
                    for (int j = 0; j < citycount; j++)
                        Bestgeti.X[j] = Population[i].X[j];
                }
            }
            Bestgeti.fitness = Evaluate(Bestgeti);
        }

        void ShuchuPopulation(int ite)
        {
            for (int i = 0; i < Pop_Size; i++)
            {
                cout << "第"<<ite<<"代种群中第"<<i + 1<<"个鲸鱼"<<"->";
                for (int j = 0; j < dimension; j++)
                {
                    if (j == dimension - 1) cout << Population[i].X[j] <<")对应的适应度为： "<<Evaluate( Population[i])<< endl;
                    else if(j==0)
                        cout <<"("<<  Population[i].X[j] << ", ";
                    else
                        cout <<  Population[i].X[j] << ", ";
                }
            }
        }
        void ShuchuBestgeti(int ite)
        {
            for (int j = 0; j < dimension; j++)
            {
                if (j == dimension - 1) std::cout << Bestgeti.X[j] << ")" << "对应的适应度为：" << Evaluate(Bestgeti) << endl;
                else if (j == 0) std::cout <<  "第"<<ite<<"代种群最优鲸鱼个体(" << Bestgeti.X[j] << ",";
                else std::cout << Bestgeti.X[j] << ",";
            }
        }
        void WOA_TSP(int popsize,double dimen,double lb,double ub,int Max_iter,string filename)
        {
            ofstream outfile;
            outfile.open("result.txt",ios::trunc);
            Init(popsize,dimen,lb,ub,Max_iter,filename);

            outfile<<"初始化鲸鱼种群如下："<<endl;
            for (int i = 0; i < Pop_Size; i++)
            {
                outfile << "初始化种群中第"<<i + 1<<"个鲸鱼"<<"->";
                for (int j = 0; j < dimension; j++)
                {
                    if (j == dimension - 1) outfile << Population[i].X[j] <<")对应的适应度为： "<<Evaluate( Population[i])<< endl;
                    else if(j==0)
                        outfile <<"("<<  Population[i].X[j] << ", ";
                    else
                        outfile <<  Population[i].X[j] << ", ";
                }
            }
            outfile<<"初始化最优的鲸鱼个体如下："<<endl;
            for (int j = 0; j < dimension; j++)
            {
                if (j == dimension - 1) outfile << Bestgeti.X[j] << ")" << "对应的适应度为：" << Evaluate(Bestgeti) << endl;
                else if (j == 0) outfile <<  "初始种群最优鲸鱼个体(" << Bestgeti.X[j] << ",";
                else outfile << Bestgeti.X[j] << ",";
            }

            int t=0;
            while (t < Itetime)
            {
                double a = 2 - t * (2 / Itetime);  // a decreases linearly fron 2 to 0
                double a2 = -1 + t * ((-1) / Itetime);  // a2 linearly dicreases from -1 to -2 to calculate t
                for (int i = 0; i < Pop_Size; i++)
                {
                    double r1 = rand() % (N + 1) / (float)(N + 1);  // r1 is a random number in [0,1]
                    double r2 = rand() % (N + 1) / (float)(N + 1);  // r2 is a random number in [0,1]
                    double A = 2 * a * r1 - a;
                    double C = 2 * r2;
                    double b = 1;
                    double l = (a2 - 1) * (rand() % (N + 1) / (float)(N + 1)) +1;
                    double p = rand() % (N + 1) / (float)(N + 1);
                    for(int j=0;j<dimension;j++)
                    {
                        if(p < 0.5)
                        {
                            if(abs(A) >= 1)
                            {
                                int rand_leader_index = rand() % dimension;
                                JingYu X_rand;
                                X_rand.Init();
                                for(int k=0;k<dimension;k++)
                                    X_rand.X[k] = Population[rand_leader_index].X[k];
                                double D_X_rand = abs(C * X_rand.X[j] - Population[i].X[j]);
                                Population[i].X[j] = int(X_rand.X[j] - A * D_X_rand);
                            }
                            else
                            {
                                double D_Leader = abs(C * Bestgeti.X[j] - Population[i].X[j]);
                                Population[i].X[j] = Bestgeti.X[j] - A * D_Leader;
                            }
                        }
                        else
                        {
                            double distance2Leader = abs(Bestgeti.X[j] - Population[i].X[j]);
                            Population[i].X[j] = int(distance2Leader * exp(b * l) * cos(l * 2 * PI) + Bestgeti.X[j]);
                        }
                    }
                }
                t = t + 1;
                for (int i = 0; i < Pop_Size; i++)
                {
                    Adjuxt_validGeti(Population[i]);
                    Population[i].fitness = Evaluate(Population[i]);
                }
                UpdateBestgeti();
                ShuchuPopulation(t);
                ShuchuBestgeti(t);
                //写到文本文件中
                for (int i = 0; i < Pop_Size; i++)
                {
                    outfile << "第"<<t<<"代种群中第"<<i + 1<<"个鲸鱼"<<"->";
                    for (int j = 0; j < dimension; j++)
                    {
                        if (j == dimension - 1) outfile << Population[i].X[j] <<")对应的适应度为： "<<Evaluate( Population[i])<< endl;
                        else if(j==0)
                            outfile <<"("<<  Population[i].X[j] << ", ";
                        else
                            outfile <<  Population[i].X[j] << ", ";
                    }
                }
                for (int j = 0; j < dimension; j++)
                {
                    if (j == dimension - 1) outfile << Bestgeti.X[j] << ")" << "对应的适应度为：" << Evaluate(Bestgeti) << endl;
                    else if (j == 0) outfile <<  "第"<<t<<"代种群最优鲸鱼个体(" << Bestgeti.X[j] << ",";
                    else outfile << Bestgeti.X[j] << ",";
                }
            }
            cout<<"****************迭代结束！****************"<<endl;
            for (int j = 0; j < dimension; j++)
            {
                if (j == dimension - 1) std::cout << Bestgeti.X[j] << ")" << "对应的适应度为：" << Evaluate(Bestgeti) << endl;
                else if (j == 0) std::cout <<  Max_iter<<"次迭代后最终得到的最优鲸鱼个体(" << Bestgeti.X[j] << ",";
                else std::cout << Bestgeti.X[j] << ",";
            }

            outfile<<"****************迭代结束！****************"<<endl;
            for (int j = 0; j < dimension; j++)
            {
                if (j == dimension - 1) outfile << Bestgeti.X[j] << ")" << "对应的适应度为：" << Evaluate(Bestgeti) << endl;
                else if (j == 0) outfile <<  Max_iter<<"次迭代后最终得到的最优鲸鱼个体(" << Bestgeti.X[j] << ",";
                else outfile << Bestgeti.X[j] << ",";
            }
            outfile.close();
        }
};

int main()
{
    srand((int)time(0));  // 产生随机种子，否则每次的随机结果都是一样
	std::cout << "****************鲸鱼优化算法求解旅行商问题！****************" << endl;
	WOA woa;
	woa.WOA_TSP(50,29,1,29,100,"D:\\program files\\CodeBlocks\\myworkspace\\WOA_TSP\\bayg29.tsp");
	return 0;
}
